File: src/policy.ts
import { parseDocument } from "yaml";
import { minimatch } from "minimatch";

export const dependabotLogin = "dependabot[bot]";
const ranks = { patch: 1, minor: 2, major: 3 } as const;
type UpdateType = keyof typeof ranks;
export interface Config {
  enabled: boolean;
  ecosystems: Record<string, UpdateType>;
  ignore: string[];
}
export interface PolicyPR {
  authorLogin: string;
  headRef: string;
  title: string;
  headCommitMessage?: string;
}
export interface Decision { merge: boolean; reason: string }

export function defaultConfig(): Config {
  return { enabled: true, ecosystems: { github_actions: "major", "*": "minor" }, ignore: [] };
}
function object(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
export function parseConfig(text: string): Config {
  const doc = parseDocument(text, { uniqueKeys: true });
  if (doc.errors.length) throw new Error(`parsing depsmate.yml: ${doc.errors[0].message}`);
  const user: unknown = doc.toJS({ maxAliasCount: 100 });
  const cfg = defaultConfig();
  if (user === null) return cfg;
  if (!object(user)) throw new Error("config must be a mapping");
  for (const key of Object.keys(user)) {
    if (!["enabled", "ecosystems", "ignore"].includes(key)) throw new Error(`unknown config key: ${key}`);
  }
  if (user.enabled != null) {
    if (typeof user.enabled !== "boolean") throw new Error("enabled must be boolean");
    cfg.enabled = user.enabled;
  }
  if (user.ecosystems != null) {
    if (!object(user.ecosystems)) throw new Error("ecosystems must be a mapping");
    for (const [eco, type] of Object.entries(user.ecosystems)) {
      if (typeof type !== "string" || !Object.hasOwn(ranks, type)) throw new Error(`invalid update type for ${eco}`);
      Object.defineProperty(cfg.ecosystems, eco, { value: type, enumerable: true, configurable: true, writable: true });
    }
  }
  if (user.ignore != null) {
    if (!Array.isArray(user.ignore) || !user.ignore.every(v => typeof v === "string")) throw new Error("ignore must be a list of strings");
    cfg.ignore = user.ignore;
  }
  return cfg;
}

export function evaluate(cfg: Config, pr: PolicyPR): Decision {
  const deny = (reason: string): Decision => ({ merge: false, reason });
  if (!cfg.enabled) return deny("disabled by .github/depsmate.yml");
  if (pr.authorLogin !== dependabotLogin) return deny(`author ${pr.authorLogin} is not ${dependabotLogin}`);
  const eco = /^dependabot\/([^/]+)/.exec(pr.headRef)?.[1];
  if (!eco) return deny(`head ref ${pr.headRef} is not a dependabot branch`);
  const maxType = Object.hasOwn(cfg.ecosystems, eco) ? cfg.ecosystems[eco] : cfg.ecosystems["*"];
  const maxRank = ranks[maxType] ?? 0;
  const message = pr.headCommitMessage ?? "";
  const names = [...message.matchAll(/^\s*-\s*dependency-name:\s*"?([^"\s]+)"?\s*$/gm)].map(m => m[1]);
  const types = [...message.matchAll(/^\s*update-type:\s*version-update:semver-(major|minor|patch)\s*$/gm)].map(m => m[1] as UpdateType);
  if (names.length !== types.length) return deny(`commit metadata lists ${names.length} dependencies but ${types.length} update types`);
  if (!names.length) {
    const bumps = [...pr.title.matchAll(/\bbumps? (\S+) from v?(\S+) to v?(\S+)/gi)];
    if (bumps.length !== 1) return deny(`${eco} bump has no metadata or single-dependency title`);
    const [, name, from, to] = bumps[0];
    const f = from.split("."), t = to.split(".");
    names.push(name);
    types.push(f[0] !== t[0] ? "major" : f.length > 1 && t.length > 1 && f[1] !== t[1] ? "minor" : "patch");
  }
  for (let i = 0; i < names.length; i++) {
    // Match a whole dependency name; wildcards do not cross path separators.
    if (cfg.ignore.some(pattern => minimatch(names[i], pattern, {
      dot: true, noglobstar: true, noext: true, nobrace: true, nonegate: true, nocomment: true,
    }))) return deny(`dependency ${names[i]} is in the ignore list`);
    if (ranks[types[i]] > maxRank) return deny(`${eco} bump of ${names[i]} is semver-${types[i]}, above allowed semver-${maxType}`);
  }
  return { merge: true, reason: `${eco} bump, all ${names.length} update(s) within semver-${maxType}` };
}

File: pkg/policy/policy.go
// Package policy decides whether a dependabot PR is safe to auto-merge.
package policy

import (
	"fmt"
	"path"
	"regexp"
	"strings"

	"sigs.k8s.io/yaml"
)

// DependabotLogin is the GitHub login dependabot authors PRs under.
const DependabotLogin = "dependabot[bot]"

// Config is the per-repo policy, read from .github/depsmate.yml on the PR's
// base branch. Zero-value fields fall back to the defaults in Default().
type Config struct {
	// Enabled turns depsmate off for the repo when set to false.
	Enabled *bool `json:"enabled"`
	// Ecosystems sets the largest semver bump auto-merged ("patch",
	// "minor", or "major") per dependabot ecosystem (the branch-name
	// form: github_actions, go_modules, npm_and_yarn, pip, ...), with
	// "*" as the fallback for unlisted ecosystems. Entries are merged
	// over the defaults {github_actions: major, "*": minor}, so listing
	// one ecosystem leaves the others at their defaults.
	Ecosystems map[string]string `json:"ecosystems"`
	// Ignore lists dependency-name globs (path.Match syntax) that are
	// never auto-merged.
	Ignore []string `json:"ignore"`
}

// Default returns the built-in policy: minor/patch library bumps, any
// github-actions bump.
func Default() Config {
	enabled := true
	return Config{
		Enabled:    &enabled,
		Ecosystems: map[string]string{"github_actions": "major", "*": "minor"},
	}
}

// ParseConfig parses .github/depsmate.yml and merges it over Default(),
// rejecting invalid values.
func ParseConfig(data []byte) (Config, error) {
	var user Config
	// Strict: unknown (typoed) keys are an error, so they fail closed
	// instead of silently applying defaults.
	if err := yaml.UnmarshalStrict(data, &user); err != nil {
		return Config{}, fmt.Errorf("parsing depsmate.yml: %w", err)
	}
	cfg := Default()
	if user.Enabled != nil {
		cfg.Enabled = user.Enabled
	}
	for eco, ut := range user.Ecosystems {
		if _, ok := updateTypeRank[ut]; !ok {
			return Config{}, fmt.Errorf("invalid update type %q for ecosystem %q", ut, eco)
		}
		cfg.Ecosystems[eco] = ut
	}
	cfg.Ignore = user.Ignore
	return cfg, nil
}

// PR carries the pull-request fields the merge policy inspects.
type PR struct {
	AuthorLogin string
	HeadRef     string // e.g. "dependabot/go_modules/github.com/foo/bar-1.2.3"
	Title       string
	// HeadCommitMessage is the full message of the PR head commit, where
	// dependabot embeds an updated-dependencies YAML block with
	// "dependency-name:" and "update-type: version-update:semver-*" lines
	// per bumped dependency.
	HeadCommitMessage string
}

// Decision is the policy outcome for one PR.
type Decision struct {
	Merge  bool
	Reason string
}

// Update is one bumped dependency extracted from dependabot metadata.
type Update struct {
	Name string
	Type string // "major", "minor", or "patch"
}

var updateTypeRank = map[string]int{"patch": 1, "minor": 2, "major": 3}

var (
	depNameRe    = regexp.MustCompile(`(?m)^\s*-\s*dependency-name:\s*"?([^"\s]+)"?\s*$`)
	updateTypeRe = regexp.MustCompile(`(?m)^\s*update-type:\s*version-update:semver-(major|minor|patch)\s*$`)
	titleBumpRe  = regexp.MustCompile(`(?i)\bbumps? (\S+) from v?(\S+) to v?(\S+)`)
)

// Evaluate applies cfg to one dependabot PR: every updated dependency must be
// outside the ignore list and within the configured max update type for the
// PR's ecosystem.
func Evaluate(cfg Config, pr PR) Decision {
	if cfg.Enabled != nil && !*cfg.Enabled {
		return Decision{Reason: "disabled by .github/depsmate.yml"}
	}
	if pr.AuthorLogin != DependabotLogin {
		return Decision{Reason: fmt.Sprintf("author %q is not %s", pr.AuthorLogin, DependabotLogin)}
	}

	eco := ecosystem(pr.HeadRef)
	if eco == "" {
		return Decision{Reason: fmt.Sprintf("head ref %q is not a dependabot branch", pr.HeadRef)}
	}
	maxType, ok := cfg.Ecosystems[eco]
	if !ok {
		// An absent "*" ranks as 0, denying everything: fail closed.
		maxType = cfg.Ecosystems["*"]
	}

	updates, err := parseUpdates(pr.HeadCommitMessage)
	if err != nil {
		return Decision{Reason: err.Error()}
	}
	if len(updates) == 0 {
		u, ok := updateFromTitle(pr.Title)
		if !ok {
			return Decision{Reason: fmt.Sprintf("%s bump has no update-type metadata and title is not a single \"bump X from A to B\"", eco)}
		}
		updates = []Update{u}
	}

	for _, u := range updates {
		if ignored(cfg.Ignore, u.Name) {
			return Decision{Reason: fmt.Sprintf("dependency %s is in the ignore list", u.Name)}
		}
		if updateTypeRank[u.Type] > updateTypeRank[maxType] {
			return Decision{Reason: fmt.Sprintf("%s bump of %s is semver-%s, above the allowed semver-%s", eco, u.Name, u.Type, maxType)}
		}
	}
	return Decision{Merge: true, Reason: fmt.Sprintf("%s bump, all %d update(s) within semver-%s", eco, len(updates), maxType)}
}

// ecosystem extracts the package ecosystem from a dependabot branch name
// ("dependabot/<ecosystem>/..."); empty when the branch is not dependabot's.
func ecosystem(headRef string) string {
	rest, ok := strings.CutPrefix(headRef, "dependabot/")
	if !ok {
		return ""
	}
	eco, _, _ := strings.Cut(rest, "/")
	return eco
}

// parseUpdates extracts the updated-dependencies metadata from the head
// commit message. A name count that disagrees with the update-type count
// means some dependency's bump kind is unknown — fail closed.
func parseUpdates(commitMessage string) ([]Update, error) {
	names := depNameRe.FindAllStringSubmatch(commitMessage, -1)
	types := updateTypeRe.FindAllStringSubmatch(commitMessage, -1)
	if len(names) == 0 && len(types) == 0 {
		return nil, nil
	}
	if len(names) != len(types) {
		return nil, fmt.Errorf("commit metadata lists %d dependencies but %d update types", len(names), len(types))
	}
	updates := make([]Update, len(names))
	for i := range names {
		updates[i] = Update{Name: names[i][1], Type: types[i][1]}
	}
	return updates, nil
}

// updateFromTitle handles PRs whose head commit carries no metadata: it
// accepts only a single-dependency "bump X from A to B" title and derives the
// update type by comparing the two versions.
func updateFromTitle(title string) (Update, bool) {
	m := titleBumpRe.FindAllStringSubmatch(title, -1)
	if len(m) != 1 {
		return Update{}, false
	}
	return Update{Name: m[0][1], Type: compareVersions(m[0][2], m[0][3])}, true
}

// compareVersions classifies a version transition by the first differing
// dot-separated component.
func compareVersions(from, to string) string {
	f := strings.Split(from, ".")
	t := strings.Split(to, ".")
	if f[0] != t[0] {
		return "major"
	}
	if len(f) > 1 && len(t) > 1 && f[1] != t[1] {
		return "minor"
	}
	return "patch"
}

func ignored(patterns []string, name string) bool {
	for _, p := range patterns {
		if ok, err := path.Match(p, name); err == nil && ok {
			return true
		}
	}
	return false
}
